server/src/lib/server.rs

```diff
@@ -350,15 +396,71 @@ async fn instance_state_put(
     Ok(HttpResponseUpdatedNoContent {})
 }
 
+async fn instance_serial_task(
+    mut detach: oneshot::Receiver<()>,
+    serial: Arc<Mutex<Serial<DispCtx, LpcUart>>>,
+    mut ws_stream: WebSocketStream<Upgraded>,
+    log: Logger,
+) -> Result<(), SerialTaskError> {
+    let mut serial = serial.lock().await;
+    let read_ready_fut = futures::future::Fuse::terminated();
+    tokio::pin!(read_ready_fut);
+
+    loop {
+        let mut output = [0u8; 4096];
+        tokio::select! {
+            // Poll in the order written
+            biased;
+
+            // It's important we always poll the detach channel first
+            // so that a constant stream of incoming/outgoing messages
+            // don't cause us to ignore a detach
+            _ = &mut detach => {
+                info!(log, "Detaching from serial console");
+                let close = CloseFrame {
+                    code: CloseCode::Policy,
+                    reason: Cow::Borrowed("serial console was detached"),
+                };
+                ws_stream.send(Message::Close(Some(close))).await?;
+                break;
+            }
+
+            // Go ahead and try to read from the serial port.
+            // Backpressure: The `read_ready_fut` precondition makes sure we only try to read
+            // from the serial port until either the buffer has filled completely, or
+            // (for partial reads) 10ms have elapsed. This prevents spinning on a
+            // serial console which is emitting single bytes at a time.
+            Ok(n) = serial.read(&mut output), if read_ready_fut.is_terminated() => {
```

Comment:
(No action necessary)

I tried to see if I could refactor `read_ready_fut` into an async block - I think it would be neat if we could do something like:

```rust
Ok(n) = async {
  read_ready_fut.await;
  serial.read(&mut output).await
} => {
```

And avoid polling `read_ready_fut` below, but I don't *think* that's possible...

Reply:
Yea, I tried something like that initially but it didn't quite work as expected. Something we can always improve later.